testutil/integration/suites/authz.go

```diff
@@ -0,0 +1,109 @@
+//go:build integration
+
+package suites
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	cosmostypes "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	// poktrollMsgTypeFormat is the format for a poktroll module's message type.
+	// The first %s is the module name, and the second %s is the message name.
+	poktrollMsgTypeFormat = "/poktroll.%s.%s"
+)
+
+var defaultAuthzGrantExpiration = time.Now().Add(time.Hour)
+
+// AuthzIntegrationSuite is an integration test suite that provides helper functions for
+// running authz grant and exec messages. It is intended to be embedded in other integration
+// test suites which are dependent on authz.
+type AuthzIntegrationSuite struct {
+	BaseIntegrationSuite
+}
+
+// RunAuthzGrantMsgForPoktrollModules creates an on-chain authz grant for the given
+// granter and grantee addresses for the specified message name in each of the poktroll
+// modules present in the integration app.
+func (s *AuthzIntegrationSuite) RunAuthzGrantMsgForPoktrollModules(
+	t *testing.T,
+	granterAddr, granteeAddr cosmostypes.AccAddress,
+	msgName string,
+	moduleNames ...string,
+) {
+	t.Helper()
+
+	var foundModuleGrants = make(map[string]int)
+	for _, moduleName := range moduleNames {
+		msgType := fmt.Sprintf(poktrollMsgTypeFormat, moduleName, msgName)
+		authorization := &authz.GenericAuthorization{Msg: msgType}
+		s.RunAuthzGrantMsg(t, granterAddr, granteeAddr, authorization)
```

Comment:
Does this enable `granteeAddr` to run any command on behalf of `granterAddr`? Seems to be the case based on [1]

**Reason I'm asking**: why, for example, would we need ring signatures if we can just let applications grant relay signing to gateways on their behalf?

This is outside the scope of this PR but would love to hear if you have ideas.



[1] https://docs.cosmos.network/v0.46/modules/authz/

Reply:
@bryanchriswhite Wdyt?

Reply:
Apologies, I missed this one when replying.

### Quick review

Rings are constructed both [on-](https://github.com/pokt-network/poktroll/blob/main/x/proof/keeper/proof_validation.go#L154) and [off-chain](https://github.com/pokt-network/poktroll/blob/main/pkg/relayer/proxy/relay_verifier.go#L27). Off-chain actors use the [`RingCache`](https://github.com/pokt-network/poktroll/blob/main/pkg/crypto/rings/cache.go#L19) while on-chain rings are use the [`RingClient`](https://github.com/pokt-network/poktroll/blob/main/pkg/crypto/rings/client.go#L28`) directly. Both on- and off-chain usage depend on on-chain state: [applications' gateway delegations](https://github.com/pokt-network/poktroll/blob/main/proto/poktroll/application/types.proto#L28).


### Considering authz as an alternative to rings

An authz alternative would:

1. **Require the development of something analogous to the `RingCache` for [authz authorizations](https://docs.cosmos.network/v0.46/modules/authz/01_concepts.html#Authorization-and-Grant).** It MUST sensitive to [`MsgGrant`](https://docs.cosmos.network/v0.46/modules/authz/03_messages.html) and [`MsgRevoke`](https://docs.cosmos.network/v0.46/modules/authz/03_messages.html#msgrevoke) events.

2. **Consolidate authorization mechanisms to use the more conventional (and built-in) one.** We would have better alignment with the toolchain and ecosystem.

3. **Give up the privacy preserving property that rings and ring signatures provide.** However, I don't think we were currently relying on, nor planning to leverage this property.